packages/playwright-core/src/server/agent/agent.ts

```diff
@@ -0,0 +1,131 @@
+/**
+ * Copyright (c) Microsoft Corporation.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import fs from 'fs';
+
+import { toolsForLoop } from './backend';
+import { debug } from '../../utilsBundle';
+import { Loop, z, zodToJsonSchema } from '../../mcpBundle';
+import { runAction } from './actionRunner';
+import * as actions from './actions';
+
+import type zod from 'zod';
+import type * as loopType from '@lowire/loop';
+
+import type { Progress } from '../progress';
+import type * as channels from '@protocol/channels';
+import type { Page } from '../page';
+
+const resultSchema = z.object({
+  actions: actions.action.array(),
+});
+
+export async function pagePerform(progress: Progress, page: Page, options: channels.PagePerformParams): Promise<void> {
+  if (await cachedPerform(progress, page, options))
+    return;
+
+  const task = `
+### Instructions
+- Perform the following actions on the page.
+- Return performed actions.
+
+### Task
+${options.task}
+  `;
+  const { actions } = await perform(progress, page, task, resultSchema, options);
```

Comment:
I am somewhat worried that `resultSchema` is pretty complex, but hopefully it's ok.